### PR TITLE
feat(pass): per-kernel windowize=True opt-in for tensor window externalization

### PR DESCRIPTION
**Pull request by Leaf-Salix**
## Summary

This is a reduced follow-up to the closed #1844 attempt. Instead of introducing
global windowing policy knobs, this PR adds one explicit IR-level opt-in:
an outlined InCore-type function (`InCore`, `AIC`, or `AIV`) may carry
`windowize = true`, and Pattern 5 in `OptimizeOrchTensors` only considers
functions with that attr.

The Python DSL currently exposes this opt-in through `pl.at(...)`:

```python
with pl.at(level=pl.Level.CORE_GROUP, windowize=True):
    ...
```

Tests, tools, or handwritten DSL/IR may also attach the same function attr
directly:

```python
@pl.function(type=pl.FunctionType.InCore, attrs={"windowize": True})
def kernel(...):
    ...
```

Only outlined InCore-type kernels whose source scope explicitly sets
`windowize=True` are eligible for local tensor window externalization. Kernels
without the attr keep the existing behavior. The transform itself is keyed on
the function attr rather than on `pl.at` specifically, so other frontends or
hand-written IR can use the same attr without adding another pass option.
After lowering, both forms are equivalent because Pattern 5 reads only the
function attr.

The intended review shape is deliberately narrow:

- No global window policy API.
- No default automatic windowization.
- No carrier parameter system.
- No broad codegen/runtime surface change.
- Windowization is reachable only through explicit per-function opt-in.

### API and IR plumbing

- Adds `windowize: bool = False` to `pl.at(...)`.
- Parses `windowize` as a boolean literal.
- Prints `windowize=True` in Python IR round-trips.
- Propagates the scope attr to outlined InCore-type function attrs.
- Also supports direct function attrs such as
  `@pl.function(..., attrs={"windowize": True})` for tests/tools that construct
  IR without going through `pl.at`.

### OptimizeOrchTensors

- Adds an optional Pattern 5 module for local tensor window externalization.
- Skips Pattern 5 entirely unless at least one InCore-type function has
  `windowize=True`.
- Rewrites only explicitly opted-in kernels.
- Uses conservative safety checks before rewriting callsites:
  - materializable tensor view types,
  - dynamic parent/window compatibility,
  - submit ABI argument limits,
  - overlapping/sibling output writers,
  - `Submit` call preservation.
- Leaves unsupported cases on the baseline path instead of partially rewriting
  them.

### Unit Tests

```bash
pytest tests/ut/ir/transforms/test_optimize_orch_tensors.py -q
```

Result:

```text
64 passed
```

Coverage includes parser/outliner propagation of `windowize=True`, Python
printer round-trip, default-off behavior, positive windowed rewrites, structural
baseline fallback checks, and submit-aware ABI/safety edge cases.

### Compile-only Smoke Test

I also ran a local compile-only smoke test using
`pypto-lib/models/qwen3/14b/prefill_fwd.py`. The model source was copied to a
temporary report directory before injecting `windowize=True`; `pypto-lib` itself
was not modified.

Generated orchestration:

```text
reports/pypto-window-externalization-v7/local_prefill_compile_check_20260625_170820/
  v7_windowize_prefill_compile_only_clean2/orchestration/prefill_fwd.cpp
```

Result summary:

```text
windowed kernels: 15
.view(...) count: 34
add_inout/add_input/add_output/add_scalar: 31/51/3/54
carrier: 0
runtime_current/barrier: 0
```

The generated C++ contains the expected `__windowed` tasks and `.view(...)`
window arguments for the annotated kernels.

### Hardware Swimlane Evidence

Hardware swimlane measurements were run on Ascend NPU with model-local
`windowize=True` annotations. These annotations are not part of this PR; they
are validation evidence that the explicit opt-in exposes narrower runtime
windows for selected kernels.

| Case | Baseline (us) | v7 windowize (us) | Improvement |
| ---- | ---: | ---: | ---: |
| `qwen3/14b/prefill_fwd.py` | 9768.78 | 5741.77 | **41.22%** |
| `deepseek/v3_2/deepseek_v3_2_decode_back.py` | 2091.66 | 1760.36 | **15.84%** |
| `deepseek/v4/hc_head.py` | 160.54 | 135.98 | **15.30%** |
| `deepseek/v3_2/deepseek_v3_2_decode_front.py` | 113589.20 | 96718.16 | **14.85%** |

`qwen3/14b/prefill_fwd.py` is the primary evidence. The local runner annotated
15 kernels:

```text
rmsnorm
q_proj
kv_proj
q_norm
k_norm
attention_writeback
out_proj
out_proj_residual
post_rmsnorm
gate_proj
up_proj
silu
down_proj
down_proj_residual
final_rmsnorm
```

All 15 produced `__windowed` clones. The generated orchestration used explicit
local views and kept the submit ABI at:

```text
add_inout/add_input/add_output/add_scalar = 31/51/3/54
```

No carrier parameters and no runtime-current barriers were produced.

`deepseek/v3_2/deepseek_v3_2_decode_back.py` is the clearest DeepSeek swimlane
example: 7 `__windowed` clones, 10 `.view(...)` calls, and visible overlap
improvement in the swimlane.

`deepseek/v4/hc_head.py` is small but clean: 3 `__windowed` clones
(`hc_head_linear`, `hc_head_reduce`, `hc_head_rms`) and a consistent runtime
reduction.

`deepseek/v3_2/deepseek_v3_2_decode_front.py` is a larger supplementary case:
21 `__windowed` clones and a clean end-to-end reduction, though its swimlane is
less visually direct than `decode_back`.

## Notes

- #1844 is closed and should be treated as the previous broader attempt, not as
  the implementation shape of this PR.

### CODE REVIEW

File: python/pypto/language/parser/ast_parser.py [hunk 1]
```diff
@@ -4165,6 +4174,9 @@ def _parse_at_scope(
             deps_kw, no_dep_args_kw, dumps_kw, optional_vars, state.allow_early_resolve, span
         )
         scope_attrs = self._append_split_slot_num_attr(scope_attrs, state.split_slot_num)
+        if state.windowize:
+            scope_attrs = list(scope_attrs) if scope_attrs else []
+            scope_attrs.append(("windowize", True))
```
[the remark this answers is not in the document]
Comment (Leaf-Salix) [reply python/pypto/language/parser/ast_parser.py:4179]: Fixed in acadc643

File: tests/ut/ir/transforms/test_optimize_orch_tensors.py [hunk 2]
```diff
@@ -19,16 +19,78 @@
...
+                func.body,
+                func.span,
+                type=func.func_type,
+                level=func.level,
+                role=func.role,
+                attrs=attrs,
+                requires_runtime_binding=func.requires_runtime_binding,
+            )
+        )
+    return ir.Program(functions, program.name, program.span)
```
[the remark this answers is not in the document]
Comment (Leaf-Salix) [reply tests/ut/ir/transforms/test_optimize_orch_tensors.py]: Fixed in acadc643

**Comment by Leaf-Salix** [comment]
Re-measured `qwen3/14b/prefill_fwd.py` with the current upstream runtime
(`95197fd7`, shape `batch=1, max_seq=128, num_layers=1`). Baseline is
default-off; optimized case explicitly annotates 15 model-local kernels with
`windowize=True`.

| Mode | Runs (us) | Avg (us) | Median (us) | C++ summary |
| ---- | ---: | ---: | ---: | --- |
| default-off | 9847.28 / 9750.12 / 9360.00 | 9652.47 | 9750.12 | `__windowed=0`, `.view=8`, `rt_submit=25` |
| explicit `windowize=True` | 6371.54 / 6434.46 / 6431.24 | 6412.41 | 6431.24 | `__windowed=15`, `.view=34`, `rt_submit=25` |

Average improvement:

```text
(9652.47 - 6412.41) / 9652.47 = 33.57%
```

Both modes keep the same submit ABI counts:
`add_inout/add_input/add_output/add_scalar = 31/51/3/54`.

Note: older evidence showed lower absolute orchestration time. The current
absolute numbers may be affected by recent runtime correctness hardening around
dependency/ring-buffer handling, especially runtime #1135; the table above is
the current-runtime comparison.

**Review by lyfne123 — APPROVED** [review]
